applicationset/controllers/applicationset_controller.go

```diff
@@ -357,6 +376,62 @@ func (r *ApplicationSetReconciler) Reconcile(ctx context.Context, req ctrl.Reque
 	}, nil
 }
 
+func (r *ApplicationSetReconciler) performReverseDeletion(ctx context.Context, logCtx *log.Entry, appset argov1alpha1.ApplicationSet, currentApps []argov1alpha1.Application) (time.Duration, error) {
+	requeueTime := 10 * time.Second
+	// get deletionOrder using current Apps
+	stepLength := len(appset.Spec.Strategy.RollingSync.Steps)
+	if stepLength == 0 {
+		logCtx.Info("no rolling sync steps found")
+	}
+	appMap := make(map[string]*argov1alpha1.Application)
+	for _, app := range currentApps {
+		appMap[app.Name] = &app
+	}
+
+	// Get Rolling Sync Step Maps
+	_, appStepMap := r.buildAppDependencyList(logCtx, appset, currentApps)
+	// reverse the AppStepMap to perform deletion
+	var reverseDeleteAppSteps []deleteInOrder
+	for appName, appStep := range appStepMap {
+		reverseDeleteAppSteps = append(reverseDeleteAppSteps, deleteInOrder{appName, stepLength - appStep - 1})
+	}
+
+	sort.Slice(reverseDeleteAppSteps, func(i, j int) bool {
+		return reverseDeleteAppSteps[i].Step < reverseDeleteAppSteps[j].Step
+	})
+
+	// TODO delete this block from logs after testing
+	logCtx.Debugf("reverse deletion steps of Appset %v", appset.Name)
+	for _, step := range reverseDeleteAppSteps {
+		logCtx.Debugf("step %v : app %v", step.Step, step.AppName)
+	}
+	for _, step := range reverseDeleteAppSteps {
+		logCtx.Infof("step %v : app %v", step.Step, step.AppName)
+		app := appMap[step.AppName]
+		retrievedApp := argov1alpha1.Application{}
+		if err := r.Get(ctx, types.NamespacedName{Name: app.Name, Namespace: app.Namespace}, &retrievedApp); err != nil {
+			if apierrors.IsNotFound(err) {
+				logCtx.Infof("application %s successfully deleted", step.AppName)
+				continue
+			}
+		}
+		// Check if the application is already being deleted - remove finalizer
+		if retrievedApp.DeletionTimestamp != nil {
+			logCtx.Infof("Deletion of the application %v has been triggered, has a deletion timestamp but not completed deletion", step.AppName)
```

Comment:
nit pick: should start with a small case. Suggestion: `application %s has been marked for deletion, but object not removed yet` wdyt ?

Reply:
done in [8cc120e](https://github.com/argoproj/argo-cd/pull/23465/commits/8cc120ee5768f6fd4d1a7eac730e6e6e2359f909)